### PR TITLE
HPCC-14931 Add ability to sign ecl standard libraries

### PR DESCRIPTION
**Pull request by Michael-Gardner**
The ability to sign modules is turned on using the cmake -DSIGN_MODULES=ON flag. With SIGN_MODULES=OFF (default) the behavior remains as it was before this PR. It is assumed that a user has a private/public key pair on the build system when using SIGN_MODULES. The current implementation uses the default private/public key in the users keychain, and packages their public key with the platform. The public key is installed under the username that gets created at installation for running our binaries (default: hpcc). This is so that the pipe->run() for the gpg --verify will pick up the correct public key.

### CODE REVIEW

**Comment by Michael-Gardner** [comment]
@richardkchapman not sure who you want to have review this?

**Comment by richardkchapman** [comment]
I will

**Comment by richardkchapman** [comment]
I was prompted multiple times for the passphrase for my key during the build process - I assume that's expected, and that we will use some sort of agent to avoid that when doing official builds?

**Comment by richardkchapman** [comment]
(very confusing on a parallel make, as there were more prompts for entering the password than first appeared)

**Comment by Michael-Gardner** [comment]
I could probably figure out a way to wrap the gpg command in an expect script and have the user enter their passphrase at build time.

**Comment by richardkchapman** [comment]
Build eventually failed with:

CMake Error at /home/rchapman/hpcc/ossd/ecllibrary/std/cmake_install.cmake:68 (FILE):
  file INSTALL cannot find "/home/rchapman/hpcc/ossd/ecllibrary/std/pub.key".

Am I expected to copy something there manually if I set this option? If so should the build process check at the start to see if it is present?

**Comment by Michael-Gardner** [comment]
no.  pub.key should be getting exported from your keychain to that location during one of the targets.  I'll look into it.  It probably have something to do with the use of a passphrase on the key.

**Comment by Michael-Gardner** [comment]
Added batch, passphrase, and no-tty options to the gpg command.  pass your passphrase into cmake with -DSIGN_MODULES_PASSPHRASE=passphrase.  I also made the generated targets from the SIGN_MODULES macro, dependent on the pub.key target if SIGN_MODULES was set.

**Comment by Michael-Gardner** [comment]
I'm also moving all the logic involved in exporting and installing the public key from the ecllibrary/std/ directory to cmake_modules/

I should have a revision for you to review within the hour.

**Comment by Michael-Gardner** [comment]
@richardkchapman ready for you to check again

**Comment by richardkchapman** [comment]
Getting a warning:

Exporting public key for eclcc signed modules to /home/rchapman/hpcc/ossd/pub.key
[  0%] [  0%] gpg: WARNING: nothing exported

and an error

CMake Error at /home/rchapman/hpcc/ossd/cmake_install.cmake:36 (FILE):
  file INSTALL cannot find "/home/rchapman/hpcc/ossd/pub.key".

File: cmake_modules/commonSetup.cmake [hunk 1]
```diff
@@ -252,6 +253,24 @@ IF ("${COMMONSETUP_DONE}" STREQUAL "")
 
   set(CMAKE_MODULE_PATH "${HPCC_SOURCE_DIR}/cmake_modules/")
 
+  if(UNIX AND SIGN_MODULES)
+    #export gpg public key used for signing to new installation
+    add_custom_command(OUTPUT ${CMAKE_BINARY_DIR}/pub.key
+      COMMAND gpg --export --output=${CMAKE_BINARY_DIR}/pub.key --batch --no-tty
+      BYPRODUCTS ${CMAKE_BINARY_DIR}/pub.key
```
Comment (richardkchapman) [inline cmake_modules/commonSetup.cmake]: This line is causing problems on my version of cmake - it ends up being treated as part of the command line.

Not sure it is needed, so may be simplest to simply remove it

Comment (richardkchapman) [inline cmake_modules/commonSetup.cmake]: Alternatively move it down a line or two - that way older CMAKE's don't treat it as part of the command line (seems to just ignore it then).

**Comment by richardkchapman** [comment]
This seems to be assuming that there is just the one public/private key installed in the build-machine's GPG system. Is that a valid assumption? Should we allow selection of the one to use?

**Comment by richardkchapman** [comment]
@Michael-Gardner Other than the comments above, seems to be working fine (I tested it with BYPRODUCTS line moved down a little).

I found a couple of minor cosmetic issues in the code signing support while testing it - Jiras opened for them.

@JamesDeFabia In order to enable the feature (preventing access to unsigned external or embedded code or programs), you need to set one or more of the following in the eclccserver options:

   <Option name="--allowsigned" value="extern"/>
   <Option name="--allowsigned" value="cpp"/>
   <Option name="--allowsigned" value="pipe"/>

**Comment by Michael-Gardner** [comment]
I've gone ahead and made some revisions based on your comments:

- Removed BYPRODUCTS line
- Export only specific KeyID if SIGN_MODULES_KEYID is defined
- Sign with specific KeyID if SIGN_MODULES_KEYID is defined

**Pull request merged**